transcom/mymove
# Support Robust Item 125A/B/C/D to Display New Fields [delivers #164869455]

## Description

This PR adds the ability to display the new robust accessorial 125 fields in the Pre-Approval Request panel. Legacy support to show 125 PARs with base quantity field will also be supported.

We're are removing the `time` field regex on the backend since we expect the input to be free form. More context here: https://defensedigitalservice.slack.com/archives/G8MKQ1AJF/p1556149181046100

## Setup

To test:
1. Login to the Office app
2. Navigate to an HHG page 
3. In the JavaScript code, **REPLACE THE SHIPMENT ID WITH THE ONE YOU'RE TESTING**
4. Run JavaScript code below in the browser console or Chrome Script snippets
5. Refresh the page to see the new robust 125 PAR

```javascript
// our test data to create a line item
var createLineItemJson = {
    "id": "a09e0457-e3f9-4e78-8bac-2cabc13f4219",
    "tariff400ng_item_id": "e27968f9-44a7-4582-af85-f4a5891120fd",
    "tariff400ng_item": {
        "code": "125A",
        "created_at": "2019-03-28T22:52:09.780Z",
        "discount_type": "HHG",
        "id": "e27968f9-44a7-4582-af85-f4a5891120fd",
        "item": "Shuttle Service 25 or less miles",
        "location": "EITHER",
        "ref_code": "NONE",
        "requires_pre_approval": true,
        "uom_1": "FR",
        "uom_2": "NONE",
        "updated_at": "2019-03-28T22:52:09.780Z"
    },
    "location": "ORIGIN",
    "reason": "Too large to be transported safely",
    "notes": "Needed for gun safe moving",
    "date": "2019-03-28",
    "time": "2358",
    "address": {
        "street_address_1": "123 Main Ave",
        "street_address_2": "Apartment 9000",
        "street_address_3": "Montmârtre",
        "city": "Anytown",
        "state": "CA",
        "postal_code": "'90210' or 'N15 3NL'",
        "country": "USA"
    },
};

// abstracting our request so we can reuse
function runFetch(uri, method, jsonBody) {
    // regex from https://developer.mozilla.org/en-US/docs/Web/API/document/cookie#Example_2_Get_a_sample_cookie_named_test2
    // get csrf token so we can POST
    var maskedCSRFToken = document.cookie.replace(/(?:(?:^|.*;\s*)masked_gorilla_csrf\s*\=\s*([^;]*).*$)|^.*$/, "$1");

    return fetch(uri, {
        "credentials": "include",
        "headers": {
            "accept": "*/*",
            "accept-language": "en-US,en;q=0.9",
            "content-type": "application/json",
            "x-csrf-token": maskedCSRFToken,
        },
        "referrerPolicy": "no-referrer-when-downgrade",
        "body": JSON.stringify(jsonBody),
        "method": method,
        "mode": "cors"
    });
}

// REPLACE SHIPMENT ID WITH THE ONE YOU'RE TESTING
var shipmentId = 'fe85b5ba-b82f-4118-b17d-2e226538bd61';
var uri = `http://officelocal:3000/api/v1/shipments/${shipmentId}/accessorials`;

runFetch(uri, "POST", createLineItemJson)
    .then(response => response.json())
    .then(res => {
        console.log("Shipment line item created:");
        console.log(res);
    })
```

## Code Review Verification Steps

* [ ] This works in [Supported Browsers](./docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/164869455) for this change

## Screenshots

Legacy and Robust 125 PARs
![image](https://user-images.githubusercontent.com/1522549/56757131-530a4900-6748-11e9-95f9-78a13b392f97.png)
